File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        if(s == null){
            return 0;
        }
        int length = s.length();
        if(length==0){
            return 0;
        }
        int index = 0;
        while(index<length && s.charAt(index)==' '){
            index++;
        }

        if(index==length){
            return 0;
        }
        int sign  = 1;


        if(s.charAt(index)=='-'){
            sign = -1;
            index++;
        }else if(s.charAt(index)=='+'){
            index++;
        }
        int result = 0;
        int threshold = Integer.MAX_VALUE/10;
        while(index<length){
            char ch = s.charAt(index);
            if(ch<'0'||ch>'9'){
                break;
            }

            if(result>threshold || (result==threshold && ch>'7')){
                return sign ==1?Integer.MAX_VALUE:Integer.MIN_VALUE;
            }
            result = result*10 + (ch-'0');
            index++;

        }return sign*result;
    }
}

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        if(not s):
            return 0
        length = len(s)

        index =0
        while index < length and s[index] == " ":
            index += 1

        if(index==length):
            return 0
        sign = 1
        if(s[index]=="-"):
            sign = -1
            index+=1
        elif(s[index]=="+"):
            sign = 1
            index+=1
        result = 0
        INT_MAX = (2**31-1)
        INT_MIN = -(2**31)
        threshold = INT_MAX//10
        while(index<length and s[index].isdigit()):
            digit = int(s[index])
            if(result>threshold or (result==threshold and digit>7)):
                return INT_MAX if sign == 1 else INT_MIN
            result = result*10+digit
            index+=1

        return sign*result